### PR TITLE
Improve Qwen3.5 recurrent cache handling

### PR DESCRIPTION
**Pull request by aleroot**
## Proposed changes

Qwen3.6 models using `model_type: qwen3_5` rely heavily on hybrid linear-attention / GatedDelta layers. Keeping the convolution state contiguous avoids carrying strided slices through decode steps, which should reduce per-token cache overhead and better match upstream MLX Python behavior.

Advancing array-cache metadata and preserving left-padding masks also improves stability for padded or batched generation paths.

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: Libraries/MLXLMCommon/KVCache.swift [hunk 1]
```diff
@@ -1157,9 +1157,17 @@ public class ArraysCache: BaseKVCache {
         leftPadding = nil
     }
 
+    /// Advance sequence-relative metadata after consuming `N` tokens.
+    public func advance(_ N: Int) {
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/KVCache.swift]: This looks ok, but the python code:

```python
    def advance(self, N):
        if self.lengths is not None:
            self.lengths -= N
        if self.left_padding is not None:
            self.left_padding -= N
```

has a `lengths` property and doesn't touch`'offset`.  Perhaps we need an effort to fully sync the python and swift versions of these KVCaches.

File: Libraries/MLXLLM/Models/Qwen35.swift [hunk 2]
```diff
@@ -287,6 +287,7 @@ final class Qwen35GatedDeltaNet: Module {
 
         if let cache {
             cache[1] = state
+            cache.advance(S)
```
Comment (davidkoski) [inline Libraries/MLXLLM/Models/Qwen35.swift:290]: FalconH1 also uses this cache -- does it need the same treatment?

Comment (davidkoski) [reply Libraries/MLXLLM/Models/Qwen35.swift:290]: Oh, actually anything that uses MambaCache is also using an ArrraysCache -- do they also need it?

Comment (aleroot) [reply Libraries/MLXLLM/Models/Qwen35.swift:290]: Yes, indeed. I have tried to cover them all, have look I hope I have not missed anything

**Comment by aleroot** [comment]
Updated rebasing on latest main , also fixed a MTP tail-budget issue by making speculative rounds budget-aware: the iterator now only drafts when the remaining output budget can hold both the accepted draft prefix and the verifier correction/bonus.

File: Libraries/MLXLMCommon/KVCache.swift [hunk 3]
```diff
@@ -93,6 +93,43 @@ extension KVCache {
...
+        }
+    }
+
+    fileprivate func finalizeArrayMetadata() {
+        if let arrays = self as? ArraysCache {
+            arrays.finalize()
+        } else if let list = self as? CacheList {
+            list.finalize()
+        }
+    }
```
Comment (davidkoski) [inline Libraries/MLXLMCommon/KVCache.swift]: I wonder if this should be done through either methods on the KVCache protocol with default (empty) implementations?  Or add a KVCacheLifecycle (or something) protocol.  I worry that new cache types could get added (we already have quite a few) and this particular check wouldn't be updated.

**Review by davidkoski — COMMENTED** [review]
Looks good.  See what you think about my question on KVCache finalize.

File: Libraries/MLXLMCommon/KVCache.swift (hunk 3 above)
Comment (aleroot) [reply Libraries/MLXLMCommon/KVCache.swift]: Agree. Moved this to KVCache lifecycle hooks with no-op defaults, made CacheList forward via the protocol, added coverage, and fixed the format issue. Thanks.

**Comment by aleroot** [comment]
Had to fix the the CI failure which was coming from `SpeculativeDecodingTests`, specifically speculative generation diverging from normal generation.

There were two issues:

1. The parameterized Swift Testing cases were running concurrently, while this suite uses shared MLX/model generation state. I marked the suite `.serialized` so the cases run one at a time.

2. The repetition/presence penalty processors were mutating logits via indexed assignment. In the speculative verification path those logits can be views into a larger multi-token logits tensor, so that could affect later verification positions. I changed those processors to return out-of-place adjusted logits using `takeAlong`/`putAlong`.

**Review by davidkoski — APPROVED** [review]
Looks good, thank you!
